### DISCUSSION TITLE
Recommended way of clearing queues

### DISCUSSION THREAD

**Discussion by add-n2x** [savonet/liquidsoap#2546] [Q&A] [answered]
I am clearing all requests in a queue channel, after switching to another channel. Just to have a clean slate, when switching back to the queue channel.

After a conversion in Liquidsoap Slack, I got the tip to implement clearing of queues this way in Liquidsoap 2:

```ocaml
def clear_queue(s) =
    q = s.queue()
    s.set_queue([])
    list.iter(request.destroy, q)
end
```
This function is called by a server function.

Now I see Liquidsoap 2 queues also have the new and native `flush_and_skip` server function, which is implemented like this:

```ocaml
"flush_and_skip", fun (_) -> try
  s.set_queue([])
  s.skip()
  "Done."
catch err do
  "Error while flushing and skipping source: #{err}"
end
```

Which way to you recommend, since the first version also destroys the requests? Are requests now automatically garbage collected, when there is nothing pointing to them?

**Comment by smimram** [comment, marked as answer]
Yes the requests are now garbage collected (https://github.com/savonet/liquidsoap/pull/2463), which means that you don't have to iterate `request.destroy` anymore. I would however advise that you still do it in case you expect your `clear_queue` function to be called intensively because there is now guarantee abouth when the destroying will be performed.

**Reply by add-n2x** [reply]
Fantastic. Thanks for the answer.

So I went for a combination of the two:

```ocaml
def clear_queue(s) =
    q = s.queue()
    s.set_queue([])
    s.skip()
    list.iter(request.destroy, q)
end
```
